File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/metrics/SpringTimer.java
package com.github.dynamicextensionsalfresco.metrics;

import org.alfresco.repo.transaction.AlfrescoTransactionSupport;
import org.alfresco.repo.transaction.TransactionListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.transaction.support.TransactionSynchronizationManager;
import org.springframework.util.StopWatch;

import javax.validation.constraints.NotNull;

/**
 * {@link Timer} based on Spring's {@link StopWatch} collects data during a transaction and logs to log4j after commit
 * at TRACE level
 */
public final class SpringTimer implements Timer {

    private final Logger logger = LoggerFactory.getLogger(SpringTimer.class);

    private final String identifier = SpringTimer.class.getPackage().getName();

    @Override
    public boolean isEnabled() {
        return logger.isTraceEnabled();
    }

    @Override
    public void start(String label) {
        if (isEnabled()) {
            StopWatch stopWatch = getStopWatch();
            if (stopWatch.isRunning()) {
                stopWatch.stop();
            }
            stopWatch.start(label);
        }
    }

    @Override
    public void stop() {
        if (isEnabled()) {
            StopWatch stopWatch = getStopWatch();
            if (stopWatch.isRunning()) {
                stopWatch.stop();
            }
        }
    }

    @NotNull
    private StopWatch getStopWatch() {
        Object stopWatch = TransactionSynchronizationManager.getResource(identifier);
        if (!(stopWatch instanceof StopWatch)) {
            stopWatch = null;
        }

        if (stopWatch == null) {
            stopWatch = new StopWatch(identifier);
            TransactionSynchronizationManager.bindResource(identifier, stopWatch);

            registerTxListener();
        }
        return (StopWatch) stopWatch;
    }

    private void registerTxListener() {
        AlfrescoTransactionSupport.bindListener(new TransactionListener() {
            @Override
            public void flush() {
            }

            @Override
            public void beforeCommit(boolean readOnly) {
            }

            @Override
            public void beforeCompletion() {
            }

            @Override
            public void afterCommit() {
                logger.trace(getStopWatch().prettyPrint());
            }

            @Override
            public void afterRollback() {
            }
        });
    }
}

File: annotations-runtime/shared/src/main/kotlin/com/github/dynamicextensionsalfresco/metrics/SpringTimer.kt
package com.github.dynamicextensionsalfresco.metrics

import org.alfresco.repo.transaction.AlfrescoTransactionSupport
import org.alfresco.repo.transaction.TransactionListener
import org.slf4j.LoggerFactory
import org.springframework.transaction.support.TransactionSynchronizationManager
import org.springframework.util.StopWatch

/**
 * [Timer] based on Spring's [StopWatch]
 * collects data during a transaction and logs to log4j after commit at TRACE level
 */
public class SpringTimer : Timer {
    val logger = LoggerFactory.getLogger(javaClass)

    val identifier = javaClass.`package`.name

    override fun isEnabled(): Boolean {
        return logger.isTraceEnabled
    }

    private val stopWatch: StopWatch
        get() {
            var stopWatch = TransactionSynchronizationManager.getResource(identifier) as? StopWatch
            if (stopWatch == null) {
                stopWatch = StopWatch(identifier)
                TransactionSynchronizationManager.bindResource(identifier, stopWatch)

                registerTxListener()
            }
            return stopWatch
        }

    private fun registerTxListener() {
        AlfrescoTransactionSupport.bindListener(object : TransactionListener {
            override fun flush() {}

            override fun beforeCompletion() {}

            override fun beforeCommit(readOnly: Boolean) {}

            override fun afterRollback() {}

            override fun afterCommit() {
                logger.trace(stopWatch.prettyPrint())
            }
        })
    }

    override fun start(label: String) {
        if (isEnabled) {
            with(stopWatch) {
                if (isRunning) {
                    stop()
                }
                start(label)
            }
        }
    }

    override fun stop() {
        if (isEnabled) {
            with(stopWatch) {
                if (isRunning) {
                    stop()
                }
            }
        }
    }
}
